### ISSUE TITLE
MesosBatchSystemTest fails on systems with only one core

### ISSUE THREAD

**Issue by hannes-ucsc** [BD2KGenomics/slugflow#99]
To reproduce, run on t2.micro or add `--resources=cpus(*):1` to the `mesos-slave` command line in jobTree.test.batchSystems.abstractBatchSystemTest.MesosBatchSystemTest.MesosSlaveThread#mesosCommand. The test will hang:

```
INFO:abstractBatchSystemTest:Waiting for master and slave processes
I0604 00:01:35.213598 2131976960 main.cpp:181] Build: 2015-03-29 03:52:41 by brew
I0604 00:01:35.214061 2131976960 main.cpp:183] Version: 0.22.0
I0604 00:01:35.213969 2131976960 main.cpp:156] Build: 2015-03-29 03:52:41 by brew
I0604 00:01:35.214365 2131976960 main.cpp:158] Version: 0.22.0
I0604 00:01:35.214457 2131976960 containerizer.cpp:110] Using isolation: posix/cpu,posix/mem
I0604 00:01:35.214534 2131976960 main.cpp:306] Starting Mesos master
I0604 00:01:35.215642 168222720 master.cpp:349] Master 20150604-000135-16777343-5050-27372 (localhost) started on 127.0.0.1:5050
W0604 00:01:35.215682 168222720 master.cpp:353] 
**************************************************
Master bound to loopback interface! Cannot communicate with remote schedulers or slaves. You might want to set '--ip' flag to a routable IP address.
**************************************************
I0604 00:01:35.215709 168222720 master.cpp:397] Master allowing unauthenticated frameworks to register
I0604 00:01:35.215718 168222720 master.cpp:402] Master allowing unauthenticated slaves to register
I0604 00:01:35.217583 168222720 master.cpp:1356] The newly elected leader is master@127.0.0.1:5050 with id 20150604-000135-16777343-5050-27372
I0604 00:01:35.217600 168222720 master.cpp:1369] Elected as the leading master!
I0604 00:01:35.217607 168222720 master.cpp:1187] Recovering from registrar
I0604 00:01:35.217648 170369024 registrar.cpp:313] Recovering registrar
I0604 00:01:35.221926 2131976960 main.cpp:200] Starting Mesos slave
I0604 00:01:35.222506 292868096 slave.cpp:174] Slave started on 1)@127.0.0.1:5051
W0604 00:01:35.222532 292868096 slave.cpp:177] 
**************************************************
Slave bound to loopback interface! Cannot communicate with remote master(s). You might want to set '--ip' flag to a routable IP address.
**************************************************
I0604 00:01:35.222729 292868096 slave.cpp:299] Slave resources: cpus(*):1; mem(*):15360; disk(*):470808; ports(*):[31000-32000]
I0604 00:01:35.222965 292868096 slave.cpp:328] Slave hostname: localhost
I0604 00:01:35.222975 292868096 slave.cpp:329] Slave checkpoint: true
I0604 00:01:35.224393 171442176 registrar.cpp:346] Successfully fetched the registry (0B) in 6.727936ms
I0604 00:01:35.224423 295550976 state.cpp:35] Recovering state from '/tmp/mesos/meta'
I0604 00:01:35.224491 171442176 registrar.cpp:445] Applied 1 operations in 22us; attempting to update the 'registry'
I0604 00:01:35.224550 292331520 status_update_manager.cpp:197] Recovering status update manager
I0604 00:01:35.224656 293941248 containerizer.cpp:307] Recovering containerizer
I0604 00:01:35.224961 294477824 slave.cpp:3778] Finished recovery
I0604 00:01:35.225935 292331520 status_update_manager.cpp:171] Pausing sending status updates
I0604 00:01:35.225949 293941248 slave.cpp:624] New master detected at master@127.0.0.1:5050
I0604 00:01:35.228368 293941248 slave.cpp:649] No credentials provided. Attempting to register without authentication
I0604 00:01:35.228387 293941248 slave.cpp:660] Detecting new master
I0604 00:01:35.229998 171442176 registrar.cpp:490] Successfully updated the 'registry' in 5.48224ms
I0604 00:01:35.230027 171442176 registrar.cpp:376] Successfully recovered registrar
I0604 00:01:35.230094 169295872 master.cpp:1214] Recovered 0 slaves from the Registry (83B) ; allowing 10mins for slaves to re-register
I0604 00:01:35.242424 293941248 slave.cpp:2141] Got registration for executor 'jobTreeExecutor' of framework 20150604-000134-16777343-5050-27368-0000 from executor(1)@127.0.0.1:61041
W0604 00:01:35.242465 293941248 slave.cpp:2167]  Shutting down executor 'jobTreeExecutor' as the framework 20150604-000134-16777343-5050-27368-0000 does not exist
W0604 00:01:35.311094 157237248 sched.cpp:1323] 
**************************************************
Scheduler driver bound to loopback interface! Cannot communicate with remote master(s). You might want to set 'LIBPROCESS_IP' environment variable to use a routable IP address.
**************************************************
DEBUG:jobTree.batchSystems.mesos:Queueing the job command: sleep 1 with job id: 0 ...
DEBUG:jobTree.batchSystems.mesos:... queued
DEBUG:jobTree.batchSystems.mesos:Queueing the job command: sleep 1 with job id: 1 ...
DEBUG:jobTree.batchSystems.mesos:... queued
I0604 00:01:35.316220 157237248 sched.cpp:157] Version: 0.22.0
I0604 00:01:35.316457 148738048 sched.cpp:254] New master detected at master@127.0.0.1:5050
I0604 00:01:35.316625 148738048 sched.cpp:264] No credentials provided. Attempting to register without authentication
I0604 00:01:35.316963 169832448 master.cpp:1574] Received registration request for framework 'JobTree Framework (Python)' at scheduler-b22b2cc8-1868-46ba-9990-deeafd0c2cb5@127.0.0.1:61031
I0604 00:01:35.317080 169832448 master.cpp:1638] Registering framework 20150604-000135-16777343-5050-27372-0000 (JobTree Framework (Python)) at scheduler-b22b2cc8-1868-46ba-9990-deeafd0c2cb5@127.0.0.1:61031
I0604 00:01:35.317230 170905600 hierarchical.hpp:321] Added framework 20150604-000135-16777343-5050-27372-0000
I0604 00:01:35.317446 149811200 sched.cpp:448] Framework registered with 20150604-000135-16777343-5050-27372-0000
DEBUG:jobTree.batchSystems.mesos:Registered with framework ID 20150604-000135-16777343-5050-27372-0000
I0604 00:01:36.072896 168222720 master.cpp:2938] Registering slave at slave(1)@127.0.0.1:5051 (localhost) with id 20150604-000135-16777343-5050-27372-S0
I0604 00:01:36.073066 170369024 registrar.cpp:445] Applied 1 operations in 29us; attempting to update the 'registry'
I0604 00:01:36.081043 170369024 registrar.cpp:490] Successfully updated the 'registry' in 7.947008ms
I0604 00:01:36.081468 169295872 master.cpp:2995] Registered slave 20150604-000135-16777343-5050-27372-S0 at slave(1)@127.0.0.1:5051 (localhost) with cpus(*):1; mem(*):15360; disk(*):470808; ports(*):[31000-32000]
I0604 00:01:36.081464 170905600 hierarchical.hpp:455] Added slave 20150604-000135-16777343-5050-27372-S0 (localhost) with cpus(*):1; mem(*):15360; disk(*):470808; ports(*):[31000-32000] (and cpus(*):1; mem(*):15360; disk(*):470808; ports(*):[31000-32000] available)
I0604 00:01:36.081678 171442176 master.cpp:3755] Sending 1 offers to framework 20150604-000135-16777343-5050-27372-0000 (JobTree Framework (Python)) at scheduler-b22b2cc8-1868-46ba-9990-deeafd0c2cb5@127.0.0.1:61031
I0604 00:01:36.081755 295014400 slave.cpp:792] Registered with master master@127.0.0.1:5050; given slave ID 20150604-000135-16777343-5050-27372-S0
I0604 00:01:36.081876 296087552 status_update_manager.cpp:178] Resuming sending status updates
DEBUG:jobTree.batchSystems.mesos:Received offer 20150604-000135-16777343-5050-27372-O0 with cpus: 1.0 and mem: 15360.0
DEBUG:jobTree.batchSystems.mesos:Preparing to launch mesos task 0 using offer 20150604-000135-16777343-5050-27372-O0...
DEBUG:jobTree.batchSystems.mesos:...launching mesos task 0
I0604 00:01:36.084967 170905600 master.cpp:2268] Processing ACCEPT call for offers: [ 20150604-000135-16777343-5050-27372-O0 ] on slave 20150604-000135-16777343-5050-27372-S0 at slave(1)@127.0.0.1:5051 (localhost) for framework 20150604-000135-16777343-5050-27372-0000 (JobTree Framework (Python)) at scheduler-b22b2cc8-1868-46ba-9990-deeafd0c2cb5@127.0.0.1:61031
I0604 00:01:36.092313 170905600 master.cpp:3331] Sending status update TASK_ERROR (UUID: 89ae08e3-75ba-4b72-abc6-a6e5ab8c6574) for task 0 of framework 20150604-000135-16777343-5050-27372-0000 'Task uses more resources cpus(*):1.1; mem(*):32 than available cpus(*):1; mem(*):15360; disk(*):470808; ports(*):[31000-32000]'
I0604 00:01:36.092484 171978752 hierarchical.hpp:648] Recovered cpus(*):1; mem(*):15360; disk(*):470808; ports(*):[31000-32000] (total allocatable: cpus(*):1; mem(*):15360; disk(*):470808; ports(*):[31000-32000]) on slave 20150604-000135-16777343-5050-27372-S0 from framework 20150604-000135-16777343-5050-27372-0000
DEBUG:jobTree.batchSystems.mesos:Task 0 is in a state TASK_ERROR
WARNING:jobTree.batchSystems.mesos:Task 0 is in unexpected state TASK_ERROR with message 'Task uses more resources cpus(*):1.1; mem(*):32 than available cpus(*):1; mem(*):15360; disk(*):470808; ports(*):[31000-32000]'
I0604 00:01:41.241669 168759296 master.cpp:3755] Sending 1 offers to framework 20150604-000135-16777343-5050-27372-0000 (JobTree Framework (Python)) at scheduler-b22b2cc8-1868-46ba-9990-deeafd0c2cb5@127.0.0.1:61031
DEBUG:jobTree.batchSystems.mesos:Received offer 20150604-000135-16777343-5050-27372-O1 with cpus: 1.0 and mem: 15360.0
DEBUG:jobTree.batchSystems.mesos:Preparing to launch mesos task 1 using offer 20150604-000135-16777343-5050-27372-O1...
DEBUG:jobTree.batchSystems.mesos:...launching mesos task 1
I0604 00:01:41.243670 169295872 master.cpp:2268] Processing ACCEPT call for offers: [ 20150604-000135-16777343-5050-27372-O1 ] on slave 20150604-000135-16777343-5050-27372-S0 at slave(1)@127.0.0.1:5051 (localhost) for framework 20150604-000135-16777343-5050-27372-0000 (JobTree Framework (Python)) at scheduler-b22b2cc8-1868-46ba-9990-deeafd0c2cb5@127.0.0.1:61031
I0604 00:01:41.249270 169295872 master.cpp:3331] Sending status update TASK_ERROR (UUID: 8f0b4992-c765-4557-b984-d42241549b05) for task 1 of framework 20150604-000135-16777343-5050-27372-0000 'Task uses more resources cpus(*):1.1; mem(*):32 than available cpus(*):1; mem(*):15360; disk(*):470808; ports(*):[31000-32000]'
I0604 00:01:41.249400 170369024 hierarchical.hpp:648] Recovered cpus(*):1; mem(*):15360; disk(*):470808; ports(*):[31000-32000] (total allocatable: cpus(*):1; mem(*):15360; disk(*):470808; ports(*):[31000-32000]) on slave 20150604-000135-16777343-5050-27372-S0 from framework 20150604-000135-16777343-5050-27372-0000
DEBUG:jobTree.batchSystems.mesos:Task 1 is in a state TASK_ERROR
WARNING:jobTree.batchSystems.mesos:Task 1 is in unexpected state TASK_ERROR with message 'Task uses more resources cpus(*):1.1; mem(*):32 than available cpus(*):1; mem(*):15360; disk(*):470808; ports(*):[31000-32000]'
I0604 00:01:46.261590 171978752 master.cpp:3755] Sending 1 offers to framework 20150604-000135-16777343-5050-27372-0000 (JobTree Framework (Python)) at scheduler-b22b2cc8-1868-46ba-9990-deeafd0c2cb5@127.0.0.1:61031
DEBUG:jobTree.batchSystems.mesos:Declining offers
WARNING:jobTree.batchSystems.mesos:No jobs to assign. Rejecting offer
I0604 00:01:46.263216 169832448 master.cpp:2268] Processing ACCEPT call for offers: [ 20150604-000135-16777343-5050-27372-O2 ] on slave 20150604-000135-16777343-5050-27372-S0 at slave(1)@127.0.0.1:5051 (localhost) for framework 20150604-000135-16777343-5050-27372-0000 (JobTree Framework (Python)) at scheduler-b22b2cc8-1868-46ba-9990-deeafd0c2cb5@127.0.0.1:61031
I0604 00:01:46.263476 171442176 hierarchical.hpp:648] Recovered cpus(*):1; mem(*):15360; disk(*):470808; ports(*):[31000-32000] (total allocatable: cpus(*):1; mem(*):15360; disk(*):470808; ports(*):[31000-32000]) on slave 20150604-000135-16777343-5050-27372-S0 from framework 20150604-000135-16777343-5050-27372-0000
I0604 00:01:51.279183 168759296 master.cpp:3755] Sending 1 offers to framework 20150604-000135-16777343-5050-27372-0000 (JobTree Framework (Python)) at scheduler-b22b2cc8-1868-46ba-9990-deeafd0c2cb5@127.0.0.1:61031
DEBUG:jobTree.batchSystems.mesos:Declining offers
WARNING:jobTree.batchSystems.mesos:No jobs to assign. Rejecting offer
I0604 00:01:51.280859 171978752 master.cpp:2268] Processing ACCEPT call for offers: [ 20150604-000135-16777343-5050-27372-O3 ] on slave 20150604-000135-16777343-5050-27372-S0 at slave(1)@127.0.0.1:5051 (localhost) for framework 20150604-000135-16777343-5050-27372-0000 (JobTree Framework (Python)) at scheduler-b22b2cc8-1868-46ba-9990-deeafd0c2cb5@127.0.0.1:61031
I0604 00:01:51.281116 170905600 hierarchical.hpp:648] Recovered cpus(*):1; mem(*):15360; disk(*):470808; ports(*):[31000-32000] (total allocatable: cpus(*):1; mem(*):15360; disk(*):470808; ports(*):[31000-32000]) on slave 20150604-000135-16777343-5050-27372-S0 from framework 20150604-000135-16777343-5050-27372-0000
I0604 00:01:56.299124 169832448 master.cpp:3755] Sending 1 offers to framework 20150604-000135-16777343-5050-27372-0000 (JobTree Framework (Python)) at scheduler-b22b2cc8-1868-46ba-9990-deeafd0c2cb5@127.0.0.1:61031
DEBUG:jobTree.batchSystems.mesos:Declining offers
WARNING:jobTree.batchSystems.mesos:No jobs to assign. Rejecting offer
I0604 00:01:56.300833 171442176 master.cpp:2268] Processing ACCEPT call for offers: [ 20150604-000135-16777343-5050-27372-O4 ] on slave 20150604-000135-16777343-5050-27372-S0 at slave(1)@127.0.0.1:5051 (localhost) for framework 20150604-000135-16777343-5050-27372-0000 (JobTree Framework (Python)) at scheduler-b22b2cc8-1868-46ba-9990-deeafd0c2cb5@127.0.0.1:61031
I0604 00:01:56.301100 168222720 hierarchical.hpp:648] Recovered cpus(*):1; mem(*):15360; disk(*):470808; ports(*):[31000-32000] (total allocatable: cpus(*):1; mem(*):15360; disk(*):470808; ports(*):[31000-32000]) on slave 20150604-000135-16777343-5050-27372-S0 from framework 20150604-000135-16777343-5050-27372-0000
I0604 00:02:01.315681 171978752 master.cpp:3755] Sending 1 offers to framework 20150604-000135-16777343-5050-27372-0000 (JobTree Framework (Python)) at scheduler-b22b2cc8-1868-46ba-9990-deeafd0c2cb5@127.0.0.1:61031
DEBUG:jobTree.batchSystems.mesos:Declining offers
WARNING:jobTree.batchSystems.mesos:No jobs to assign. Rejecting offer
I0604 00:02:01.316866 170369024 master.cpp:2268] Processing ACCEPT call for offers: [ 20150604-000135-16777343-5050-27372-O5 ] on slave 20150604-000135-16777343-5050-27372-S0 at slave(1)@127.0.0.1:5051 (localhost) for framework 20150604-000135-16777343-5050-27372-0000 (JobTree Framework (Python)) at scheduler-b22b2cc8-1868-46ba-9990-deeafd0c2cb5@127.0.0.1:61031
I0604 00:02:01.317046 169832448 hierarchical.hpp:648] Recovered cpus(*):1; mem(*):15360; disk(*):470808; ports(*):[31000-32000] (total allocatable: cpus(*):1; mem(*):15360; disk(*):470808; ports(*):[31000-32000]) on slave 20150604-000135-16777343-5050-27372-S0 from framework 20150604-000135-16777343-5050-27372-0000
Traceback (most recent call last):

  File "/Applications/PyCharm.app/Contents/helpers/pycharm/utrunner.py", line 155, in <module>
    TeamcityTestRunner().run(all, **options)
  File "/Applications/PyCharm.app/Contents/helpers/pycharm/tcunittest.py", line 249, in run
    test(result)
  File "/usr/local/Cellar/python/2.7.9/Frameworks/Python.framework/Versions/2.7/lib/python2.7/unittest/suite.py", line 70, in __call__
    return self.run(*args, **kwds)
  File "/usr/local/Cellar/python/2.7.9/Frameworks/Python.framework/Versions/2.7/lib/python2.7/unittest/suite.py", line 108, in run
    test(result)
  File "/usr/local/Cellar/python/2.7.9/Frameworks/Python.framework/Versions/2.7/lib/python2.7/unittest/case.py", line 393, in __call__
    return self.run(*args, **kwds)
  File "/usr/local/Cellar/python/2.7.9/Frameworks/Python.framework/Versions/2.7/lib/python2.7/unittest/case.py", line 329, in run
    testMethod()
  File "/Users/hannes/workspace/bd2k/jobTree/src/jobTree/test/batchSystems/abstractBatchSystemTest.py", line 101, in testGetUpdatedJob

    self.wait_for_jobs(wait_for_completion=True)
  File "/Users/hannes/workspace/bd2k/jobTree/src/jobTree/test/batchSystems/abstractBatchSystemTest.py", line 136, in wait_for_jobs
    time.sleep(0.1)
KeyboardInterrupt
INFO:abstractBatchSystemTest:Exiting MesosMasterThread
INFO:abstractBatchSystemTest:Exiting MesosSlaveThread

Process finished with exit code 1
``

**Comment by cket** [comment]
This seems to be  caused by the mesosbatchsystem accepting an offer that was too small to actually run the task. This is strange, since I have run mesosbatchsystem on t2.micro before with tasks that are too large to ever run and they accepted those small offers with 0 tasks, and sat silently waiting for a larger offer to actually run tasks.

 Accepting offers without any tasks wasn't ideal so I changed the behavior to reject offers like that and log the situation, but I never saw any behavior like this. And I did test those changes, and it behaved as expected: reject too small offers and log that the offer was too small for the task.

**Comment by cket** [comment]
I replicated this error with the MesosBatchSystemTest by passing --resources=cpus:1, but when I try the same with testMesos and stressTest, I get the expected, proper, offer rejection and logging behavior. This is very strange.

**Comment by hannes-ucsc** [comment]
You just volunteered ;-)

Thanks for looking into this.

**Comment by hannes-ucsc** [comment]
Some more data points: With the `cpu=1.0` in following line in testGetUpdatedJob()
```
self.batchSystem.issueJob(jobCommand, memory=10, cpu=1.0 )
```
the test hangs, with `cpu=0.5` it fails an assertion `KeyError: (1, 1)` and with `cpu=0.4` it succeeds.

It's almost as if the slave reserves 0.1 for itself, meaning that concurrent tasks with cpu less than 0.45 will run concurrently. The failed assertion in the 0.5 case seems to result from Mesos killing one of the two tasks.

Can you reproduce these findings?

**Comment by hannes-ucsc** [comment]
Also note my first attempt at fixing this:

e983bd783580e8d9ad69bfd9a0e28cca04d95a4a

It did fix hangs in other tests of the same class but I overlooked testGetUpdatedJob(). We could of course, just change cpu to 0.1 and the tests will pass but I don't feel comfortable with that until we fully understand what is going on.

**Comment by cket** [comment]
The executor does reserve 0.1 cpu for itself, although i've never seen this behavior before. I succeeded in replicating this in testMesos when I changed the first task to 1.0 cpu. So what is happening is the offer is large enough for the task, but not large enough for the task and executor. Once the task is accepted, it seems like mesos combines these two resource requirements and counts them as part of the same task. That is why although the cpu req for the task is set to 1 in MesosBatchSystemTest, the log shows it as 1.1 cpu requirement. This behavior is only seen on the first task, since the executor is still alive for the subsequent offers and we only get 0.9 cpus offered. 

I think there are two ways to fix this: remove the executor resource requirements, or test if each task is large enough for the executor+the task. 

I am in favor of removing the executor's requirements. Mesos was giving a warning when it wasn't set, saying that it was recommended that I do so and that subsequent version of mesos would require it. However, I don't see any advantage of doing so, especially since the minimum executor resource req was 0.1 cpu, which seems fairly large for such a simple process. A possible downside of this is if we choose to set a hard limit on task cpu usage, which we can enable with mesos, is the first task could be throttled by mesos since mesos apparently counts the executor's resource usage as part of that first task. Hopefully mesos wouldn't do this, but the logs above indicate that mesos may bundle the executor with the first task. 

The alternative is a little harder, since we don't have an easy way of knowing if an executor is already running in an offer so we may end up double counting executor requirements and rejecting offers that we actually could run a task on (ex: offer with 0.9 cpu from a slave with an already running executor. We have a task that uses 0.9 cpu, but if we blindly test if the offer is large enough for the task + another executor, it would not run the task on that offer)

**Comment by hannes-ucsc** [comment]
Let's dig a little deeper before we make a decision. Why is one task killed when cpu=0.5? Why is the executor CPU requirement only effective  for the first task on that executor?

**Comment by cket** [comment]
The CPU requirement is only effective on the first offer on the slave because on subsequent offers the executor is still alive, and already taken in to account. We see something like this: First offer: 1.0 cpu, xxx mem. We launch tasks on this executor and the executor is started. Second offer: 0.9 cpu, xxx-32 mem. 

This is because the executor's resource requirements are set at the minimum of 0.1 cpu and 32 mb, and the slave knows that this process is still alive so doesn't offer those resources to our framework. 

As for only one task being killed when cpu=0.5, it seems like mesos knows that the sum of these tasks is above the slaves resources, and kills off tasks until the rest can be run without a problem. Remember, in the test we are launching those two tasks at the same time on the same first offer, so the executor resources are not yet taken into account.

**Comment by cket** [comment]
http://www.mail-archive.com/user@mesos.apache.org/msg03597.html

We are supposed to keep track of whether a slave has an executor or not. I think we should just remove the executor resource requirements

**Comment by hannes-ucsc** [comment]
The second response on that thread makes more sense. And I agree we should set executor cpu requirements to zero.

**Comment by hannes-ucsc** [comment]
Be sure to undo my workaround in the unit tests.

**Comment by hannes-ucsc** [comment]
Follow-up in #108

**Comment by cket** [comment]
Closed by 9d51fdb530ee4a0a4e4cc62f27c7bd8756982608
